### PR TITLE
Add XDG support to key bindings, snippets, themes, and file templates

### PR DESCRIPTION
**Pull request by jmcphers**
This change continues the work started in https://github.com/rstudio/rstudio/pull/5069, continuing to further the 1.3 goal of (a) having all configuration data in a single, configurable folder and (b) supporting global customization of all settings.

### Details

Prior to this change, these files were stored in hard-coded (and occasionally inconsistent) locations in the user's home dir:

- Themes and keybindings were in `~/.R/rstudio`
- Snippets were just in `~/.R`
- File templates were in a built-in, read-only folder

Each of these has undergone similar work:

- All of the resources are now stored in `~/.config/rstudio` (configurable)
- For interoperability with 1.2 and prior, we support both the old and new folders. If you have any of these resources in the legacy folder location, we will continue to read from that folder. However, new content you add will be only be placed in the new folder. 
- The system folder is now read in addition to the user folder, making all of these features admin-extensible (mostly for the first time). Merging or overriding is done as appropriate. 

As an example, if a user has done some customization to RStudio, here's what their `~/.config/rstudio` folder might look like: it contains their preferences, key bindings, snippets, templates, and themes, all in plain, user-friendly text files.

```
 .
├── keybindings
│   ├── addins.json
│   ├── editor_bindings.json
│   └── rstudio_bindings.json
├── rstudio-prefs.json
├── snippets
│   └── r.snippets
├── templates
│   ├── default.R
│   ├── default.Rmd
│   ├── default.cpp
│   ├── default.js
│   └── notebook.Rmd
└── themes
    └── Brogrammer.rstheme
```

### Templates

The `templates` folder deserves some special mention since it was not previously customizable. You can now define the default content for *any* file type. For example, to set the default content for R scripts (what you see in the editor buffer when you click File -> New R Script), you can edit:

    ~/.config/rstudio/templates/default.R

Some file types, such as R Markdown, have non-default templates. These are also customizable via specially named files; for example `default.Rmd` sets the content for blank R Markdown documents, but `notebook.Rmd` sets the content for blank R Markdown notebooks.

The admin guide will document the full list of special templates.

### CODE REVIEW

**Comment by kevinushey** [comment]
> The templates folder deserves some special mention since it was not previously customizable. You can now define the default content for any file type. For example, to set the default content for R scripts (what you see in the editor buffer when you click File -> New R Script), you can edit:
> 
>     ~/.config/rstudio/templates/default.R
> 
> Some file types, such as R Markdown, have non-default templates. These are also customizable via specially named files; for example default.Rmd sets the content for blank R Markdown documents, but notebook.Rmd sets the content for blank R Markdown notebooks.

YAGNI might apply, but what do you think about allowing for multiple user-defined templates for each filetype? For example, the user might want:

    ~/.config/rstudio/templates/R/shiny-template.R
    ~/.config/rstudio/templates/R/bookdown-template.R

Ie, different templates for different single-file purposes. For now we'd only accept one template, but some theoretical day in the future this would allow us to e.g. provide a list of templates for the user to choose from (e.g. New -> From Template...)

File: src/cpp/session/modules/SessionSnippets.cpp [hunk 1]
```diff
@@ -107,33 +113,60 @@ bool isSnippetFilePath(const FilePath& filePath,
...
+   // Add system-level snippets files
+   dirs.push_back(core::system::xdg::systemConfigDir().complete("snippets"));
+
+   // Add snippets files from older RStudio
+   dirs.push_back(getLegacySnippetsDir());
+
+   // Add user snippets files
+   dirs.push_back(getSnippetsDir());
+
+   for (const auto snippetsDir: dirs)
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionSnippets.cpp]: nit: `const auto&`; prefer ref?

**Comment by jmcphers** [comment]
>  but what do you think about allowing for multiple user-defined templates for each filetype?

I think people would find it useful, but I don't think we need to build it now as it'd be meaningless without the UI. Is there something in the existing design that you think precludes us from taking this approach in the future?

**Comment by kevinushey** [comment]
> Is there something in the existing design that you think precludes us from taking this approach in the future?

I don't think so -- I guess at worst we might have to migrate the default template files into filetype-specific subfolders, but that wouldn't be too bad. (I just wanted to double-check that we're at least leaving that door open; not suggesting we should take that work on now)

**Comment by jmcphers** [comment]
> we might have to migrate the default template files into filetype-specific subfolders

I don't think we'd even need to do that! The file extension is sufficient:

```
~/.config/rstudio/templates/shiny-template.R
~/.config/rstudio/templates/bookdown-template.R
```

File: src/cpp/session/modules/SessionSnippets.cpp (hunk 1 above)
Comment (jmcphers) [reply src/cpp/session/modules/SessionSnippets.cpp]: good idea, done!

**Comment by samgreenbergBTO** [comment]
Running R Studio Version 1.3.391 on OSX, my ~/.config/rstudio folder only contains rstudio-prefs.json and /dictionaries. Any idea why I'm not seeing /templates?

**Comment by gtritchie** [comment]
> Running R Studio Version 1.3.391 on OSX, my ~/.config/rstudio folder only contains rstudio-prefs.json and /dictionaries. Any idea why I'm not seeing /templates?

@jmcphers can confirm, but my understanding is that the templates folder doesn't exist unless you create it and add your own default file templates to it.

**Comment by jmcphers** [comment]
@samgreenbergBTO @gtritchie is correct. The `~/.config` folder contains only your own content (no RStudio defaults), so you'll need to create the `templates` folder yourself if you want to try this out.

**Comment by samgreenbergBTO** [comment]
Ok, I got it working. It works for making a new R Notebook if the template is named default.Rmd or notebook.Rmd but doesn't seem to work for R Markdown. Not really a big problem since they're functionally identical but is there a specific name for markdown template? I tried markdown.Rmd.

**Pull request merged**